### ISSUE TITLE
Numerical Stability Concerns

### ISSUE THREAD

**Issue by VHellendoorn** [HazyResearch/flash-attention#16]
Hi, this seems like a super promising direction for making large language model training more scalable and accessible. However, as was also noted in #10, there seems to be a slight, but [potentially significant](https://github.com/lucidrains/memory-efficient-attention-pytorch/issues/2#issuecomment-1157529760) numerical mismatch with the reference Pytorch attention implementation. A quick check of the computed values with a 768-dimension, 4,096 sequence length setup yields: 
```
plain_fwd = attention_ref(qkv, attention_mask_bool, dropout_p, causal=causal)
flash_fwd = flash_attn_func(qkv[0], cu_seqlens, dropout_p, max_seqlen_in_batch, causal=causal)
delta = plain_fwd - flash_fwd
total_value = lambda val: val.abs().sum().item()
print(f'Total Activations: plain={total_value(plain_fwd):.2f}, flash={total_value(flash_fwd):.2f}, delta: {total_value(delta):.2f}')
# Total Activations: plain=65408.00, flash=65408.00, delta: 7.32
```
That is, while both functions achieve the same total value, small differences sum up to a meaningful delta. An inspection of the difference shows that about 86% of the values in `delta` are "true" zeros, with the remainder being off by about 1.7e-5 on average. Locations where the delta was non-zero were not particularly different in absolute activation.

Note that I removed all padding and replaced `qkv_unpad` with `qkv[0]` to make these more comparable -- using `qkv_unpad` even without any padding tokens present cause the `delta` to grow to 93.31 and the total activations to diverge. I also disabled dropout to ensure this is not due to randomness in the function itself.

Do you have any leads on where this instability might originate and if it can be remedied? If it can be fixed, this optimization could be really impactful.

-Vincent

**Comment by tridao** [comment]
Thanks for the investigation, this is great!

I think that amount of numerical error might be expected since we're using [fp16](https://en.wikipedia.org/wiki/Half-precision_floating-point_format#Precision_limitations).
This topic of numerical error is super interesting. I'd love to understand your use case better!

More details: we probably want to compare the numerical error of the Pytorch implementation in fp16 - Pytorch implementation in fp32, and Flash (in fp16) - Pytorch implementation in fp32.
They have similar max absolute error, and FlashAttention has slightly higher relative error.
This could be because of the difference in the order of operations.
Btw you can cast the delta to float so that summation doesn't exceed 64K (which is the limit of fp16).
```
ref_fp32_fwd = attention_ref(qkv, attention_mask_bool, dropout_p, causal=causal, upcast=True)
ref_fwd = attention_ref(qkv, attention_mask_bool, dropout_p, causal=causal)
flash_fwd = flash_attn_func(qkv_unpad, cu_seqlens, dropout_p, max_seqlen_in_batch, causal=causal)
flash_fwd = rearrange(flash_fwd, '(b s) h d -> b s h d', b = batch_size)
delta_ref = ref_fwd - ref_fp32_fwd
delta_flash = flash_fwd - ref_fp32_fwd
total_value = lambda val: val.float().abs().sum().item()
print(f'Reference fp16 max absolute error = {delta_ref.abs().max().item()}')
print(f'Reference fp16 mean relative error = {total_value(delta_ref) / total_value(ref_fp32_fwd)}')
print(f'Flash max absolute error = {delta_flash.abs().max().item()}')
print(f'Flash mean relative error = {total_value(delta_flash) / total_value(ref_fp32_fwd)}')

# Reference fp16 max absolute error = 6.103515625e-05
# Reference fp16 mean relative error = 7.466496773477096e-05
# Flash max absolute error = 6.103515625e-05
# Flash mean relative error = 0.00011073935366785967
```
As another example of order of operations causing numerical differences, if you simply multiply the result by 3.0 then divide it by 3.0, you get a numerical difference of the same magnitude.
```
delta_mul = ref_fwd - (ref_fwd * 3.0) / 3.0
print(f'Multiply fp16 max absolute error = {delta_mul.abs().max().item()}')
print(f'Multiply fp16 mean relative error = {total_value(delta_mul) / total_value(ref_fwd)}')
# Multiply fp16 max absolute error = 6.103515625e-05
# Multiply fp16 mean relative error = 8.713470482512265e-05
```

**Comment by tridao** [comment]
Btw we're working on implementing FlashAttention for [bf16](https://en.wikipedia.org/wiki/Bfloat16_floating-point_format) format. The folk wisdom seems to be that bf16 training is more stable than fp16 training in general.

**Comment by stephenroller** [comment]
> Btw we're working on implementing FlashAttention for [bf16](https://en.wikipedia.org/wiki/Bfloat16_floating-point_format) format. The folk wisdom seems to be that bf16 training is more stable than fp16 training in general.

+1 that would be great

**Comment by VHellendoorn** [comment]
Hi, thanks a ton for getting back to me! I appreciate the further analysis; it sounds like you are saying that these very minor deltas are expected given the low resolution of 16-bit floats (and the 64K total makes sense now, didn't think of that). The difference between `qkv[0]` and `qkv_unpad` must be a similar reason: despite `x` and `x_unpad` being identical (modulo batch shape), their results after being passed to `Wqkv` diverge by a similarly small amount.

I guess the only other risk would be if these kinds of small deltas are more prone to occurring in large attention values, since the softmax would then amplify such differences, which can even cause overflows. I haven't noticed a consistent skew like that so far, but it might still explain the instabilities that others noticed. bf16 would go a long way towards addressing that risk since it uses a _much_ larger exponent. Looking forward to seeing that support!

For some contenxt: I am considering training a large language model like [PolyCoder](https://github.com/VHellendoorn/Code-LMs) with this attention implementation to see if it will allow better scaling on limited resources. Especially the vastly reduced memory footprint should make a big difference by allowing larger batch sizes. I was trying to see if someone else had done something similar (e.g., #12 asks for the same) and found the issue by lucidrains that mentions that it seems like there are numerical instabilities that might complicate training. Hence this issue checking on whether/to what extent those could pose problems.

Thanks again!
-Vincent

**Comment by tridao** [comment]
bf16 is now supported.

**Comment by lucidrains** [comment]
@tridao tremendous! eager to hear if this solves the issue!

**Comment by VHellendoorn** [comment]
Same, thanks for adding that! @lucidrains do you still have the GPT-2 wrapper with this attention that you mentioned in #10? It'd be great to see if training is more stable now; happy to run it on my machine if you can point me to the code.

**Comment by lucidrains** [comment]
@VHellendoorn Tri indicated he has some experiments in progress https://github.com/HazyResearch/flash-attention/commit/5b838a8bef78186196244a4156ec35bbb58c337d#commitcomment-77949422 was going to touch base with him early next week

**Comment by lucidrains** [comment]
@VHellendoorn that figure actually came from the paper itself

**Comment by VHellendoorn** [comment]
Ah, I see! Thanks, I'll see if I can put something together too.

**Comment by lucidrains** [comment]
@VHellendoorn same, i'll do a small run with enwik8 later tonight, as a litmus test

**Comment by lucidrains** [comment]
all things said and done, if Tri is able to smooth out all the issues, i owe him and his team a dinner. Markus Rabe too

**Comment by VHellendoorn** [comment]
We all will. Let me know if you're ever in Pittsburgh, Tri! :) I'm definitely very excited about seeing this [get merged into Apex](https://github.com/NVIDIA/apex/issues/1395). That would help a lot of people very quickly.
